### PR TITLE
C++: Add missing parent scope cases

### PR DESCRIPTION
**Pull request by jketema**
This is mostly used to detect the shadowing of variable declarations.

DCA shows no changes. This does break some coding standards tests. I'll update the test results there after this is merged, and open an issue that informs the coding standards people of this. I briefly looked at fixing the query, but (a) it's not clearly documented what the query should do, and (b) this requires digging through about 400 lines of query code, which I find very hard to follow.

#### All query authors

- [ ] A change note is added if necessary. See [the documentation](https://github.com/github/codeql/blob/main/docs/change-notes.md) in this repository.
- [ ] All new queries have appropriate `.qhelp`. See [the documentation](https://github.com/github/codeql/blob/main/docs/query-help-style-guide.md) in this repository.
- [ ] QL tests are added if necessary. See [Testing custom queries](https://docs.github.com/en/code-security/codeql-cli/using-the-advanced-functionality-of-the-codeql-cli/testing-custom-queries) in the GitHub documentation.
- [ ] New and changed queries have correct query metadata. See [the documentation](https://github.com/github/codeql/blob/main/docs/query-metadata-style-guide.md) in this repository.

#### Internal query authors only

- [ ] Autofixes generated based on these changes are valid, only needed if this PR makes significant changes to `.ql`, `.qll`, or `.qhelp` files. See [the documentation](https://github.com/github/codeql-team/blob/main/docs/best-practices/validating-autofix-for-query-changes.md) (internal access required).
- [ ] Changes are validated [at scale](https://github.com/github/codeql-dca/) (internal access required).
- [ ] Adding a new query? Consider also [adding the query to autofix](https://github.com/github/codeml-autofix/blob/main/docs/updating-query-support.md#adding-a-new-query-to-the-query-suite).

### CODE REVIEW

File: cpp/ql/test/library-tests/scopes/parents/parents.expected [hunk 1]
```diff
@@ -14,7 +17,11 @@
 | 1 | parents.cpp:4:10:4:10 | f | parents.cpp:4:19:13:5 | { ... } |
 | 1 | parents.cpp:4:19:13:5 | { ... } | parents.cpp:5:11:5:11 | j |
 | 1 | parents.cpp:4:19:13:5 | { ... } | parents.cpp:6:11:10:7 | { ... } |
+| 1 | parents.cpp:4:19:13:5 | { ... } | parents.cpp:11:18:11:18 | e |
```
Comment (geoffw0) [inline cpp/ql/test/library-tests/scopes/parents/parents.expected:20]: It feels like the parent of `e` should be something associated with the catch block on line 11, rather than the entire function.  I can see that this wouldn't fit the current definition of what `getParentScope` can return.  And what you have is probably good enough for what `getParentScope` is used for.

**Review by geoffw0 — APPROVED** [review]

**Comment by geoffw0** [comment]
> I briefly looked at fixing the query, but (a) it's not clearly documented what the query should do, and (b) this requires digging through about 400 lines of query code, which I find very hard to follow.

Which query is this?

File: cpp/ql/test/library-tests/scopes/parents/parents.expected (hunk 1 above)
Comment (jketema) [reply cpp/ql/test/library-tests/scopes/parents/parents.expected:20]: Yeah, agreed. I struggled a bit with that one initially. What I do now seems to be most in line with all the other case that are there, and what we want when looking for shadowing.

**Comment by jketema** [comment]
This query: https://github.com/github/codeql-coding-standards/blob/b476450bfb3bf1801c96159580ec66a577ead214/cpp/common/src/codingstandards/cpp/rules/identifierhidden/IdentifierHidden.qll

**Comment by jketema** [comment]
I've reported the issue on the coding standards repo. I assume they'll have a look once they upgrade to CodeQL 2.19.3 or later.

File: cpp/ql/lib/semmle/code/cpp/Element.qll [hunk 2]
```diff
@@ -138,31 +138,40 @@ class Element extends ElementBase {
...
       this = v and
       exists(DeclStmt ds | ds.getADeclaration() = v and result = ds.getParent())
     )
     or
-    exists(Parameter p | this = p and result = p.getFunction())
+    exists(Parameter p |
+      this = p and
+      (
+        result = p.getFunction() or
+        result = p.getCatchBlock().getParent().(Handler).getParent().(TryStmt).getParent() or
```
Comment (lcartey) [inline cpp/ql/lib/semmle/code/cpp/Element.qll:158]: @jketema Could you explain the rationale behind making the parent scope of a catch block parameter the parent scope of the `TryStmt`? Surely it should be the catch block itself?

This is what is breaking the Coding Standards `IdentifierHiding.ql` query. That query tries to find pairs of identifiers where the second hides the first. By considering to the parameter of a catch block to have a parent scope of the parent of the `TryStmt` we erroneously report cases like this:
```cpp
  {
    int i; // COMPLIANT
  }

  try {

  } catch (int i) { // COMPLIANT
  }
```

Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/Element.qll:158]: > Could you explain the rationale behind making the parent scope of a catch block parameter the parent scope of the TryStmt? Surely it should be the catch block itself?

This seemed to be most consistent with the other cases here. For example, for functions the parent scope is the whole function, and not the statement block that forms the body of the function. The latter would correspond with the catch block in the case of a try-catch.

Comment (lcartey) [reply cpp/ql/lib/semmle/code/cpp/Element.qll:158]: > For example, for functions the parent scope is the whole function

Just checking, but I presume you mean "for parameters the parent scope is the whole function"?

If so, I think it would be reasonable on that basis to report the `Handler` of the `CatchBlock` as the parent scope of the parameter. But reporting the _parent_ of the `TryStmt` as the "scope" erroneously puts it in the same scope as other variables declared outside the `TryStmt`. Given the main purpose of this predicate is to support identifier hiding queries, this seems like it's not ideal.

Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/Element.qll:158]: The problem is that `Handler` is not exposed anywhere else in the QL API, and it's only really visible in the AST, so that's far from ideal either.
